applications/ParticleMechanicsApplication/custom_constitutive/johnson_cook_thermal_plastic_3D_law.cpp

```diff
@@ -0,0 +1,476 @@
+		const Matrix strain_increment_deviatoric = strain_increment - CalculateMatrixTrace(strain_increment) / 3.0 * identity;
+		const Matrix stress_deviatoric_old = stress_old - CalculateMatrixTrace(stress_old) / 3.0 * identity;
+
+		// Calculate trial (predicted) j2 stress
+		double stress_hydrostatic_new = CalculateMatrixTrace(stress_old) / 3.0 + bulk_modulus_K * CalculateMatrixTrace(strain_increment);
+		Matrix stress_deviatoric_trial = stress_deviatoric_old + 2.0 * shear_modulus_G * strain_increment_deviatoric;
+		const double j2_stress_trial = std::sqrt(3.0 / 2.0 * CalculateMatrixDoubleContraction(stress_deviatoric_trial));
+
+		// Declare deviatoric stress matrix to be used later
+		Matrix stress_deviatoric_converged = (GetStrainSize() == 3) ? Matrix(2, 2) : Matrix(3, 3);
+
+		// Assume current yield stress is the same as the old (elastic predictor)
+		double yield_stress = mYieldStressOld;
+
+		if (j2_stress_trial > yield_stress && yield_stress/mYieldStressVirgin > yield_stress_failure_ratio)
+		{
+			// Newton raphson setup
+			double gamma = mGammaOld;
+			double gamma_min = 0.0;
+			double gamma_max = j2_stress_trial / GetSqrt6() / shear_modulus_G;
+			bool is_converged = false;
+			const SizeType iteration_limit = 100;
+			IndexType iteration = 0;
+			const double tolerance_delta_gamma = 1e-9;
+			double yield_function, yield_function_gradient, dYield_dGamma,
+				delta_gamma, predicted_eps, predicted_eps_rate,
+				predicted_temperature;
+
+			// Flow direction
+			Matrix flow_direction_normalized = stress_deviatoric_trial / (j2_stress_trial / std::sqrt(3.0 / 2.0));
+
+			// Initial prediction of quantities
+			predicted_eps = mEquivalentPlasticStrainOld + GetSqrt23() * gamma; // eps = equivalent plastic strain
+			predicted_eps_rate = GetSqrt23() * gamma / CurrentProcessInfo[DELTA_TIME];
+			predicted_temperature = mTemperatureOld;
+			if (MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] > 0.0) {
+				predicted_temperature += MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] / GetSqrt6() / MaterialProperties[DENSITY]
+					/ MaterialProperties[SPECIFIC_HEAT] * (yield_stress + mYieldStressOld) * gamma;
+			}
+
+			// Newton Raphson return mapping loop
+			while (!is_converged)
+			{
+				// Calculate predicted yield stress
+				yield_stress = CalculateHardenedYieldStress(MaterialProperties, predicted_eps, predicted_eps_rate, predicted_temperature);
+
+				// Compute yield function and derivative
+				yield_function = j2_stress_trial - GetSqrt6() * shear_modulus_G * gamma - yield_stress;
+				if (yield_function < 0.0) gamma_max = gamma;
+				else gamma_min = gamma;
+				dYield_dGamma = CalculatePlasticStrainDerivative(MaterialProperties, predicted_eps, predicted_eps_rate, predicted_temperature);
+				dYield_dGamma += CalculatePlasticStrainRateDerivative(
+					MaterialProperties, predicted_eps, predicted_eps_rate, predicted_temperature)/ CurrentProcessInfo[DELTA_TIME];
+				if (MaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] > 0.0) {
```

Comment:
Can we check this at the beginning at the initialization, instead of every time step? MaterialProperties is constant, so the value is fixed as input.

Reply:
I agree it's constant throughout the simulation. In this case it is being used as a flag to indicate whether to consider thermal softening or not. The alternative would be created a member bool mIsThermal during initializeMaterial(), but the current implementation achieves this without creating an extra bool.